The use of `ROOT_OUT` in the top-level `Makefile` is broken

Introduction
----------------

The top-level `Makefile` sets an `ROOT_OUT` variable that by default is `build`. It *should* be possible to use it to build the various components in a different build directory but that mechanism does not work.

How to reproduce:
------------------------
Try to set the `ROOT_OUT` variable to a different value, e..g:
```
make ROOT_OUT=build-test
```

The build will fail as follows (only the last lines are included here):
```
[ ! -e build-test/hypervisor/arch/x86/vmx_asm.o ] && mkdir -p build-test/hypervisor/arch/x86/; \
cc  -Iinclude  -Iinclude/lib  -Iinclude/lib/crypto  -Iinclude/common  -Iinclude/arch/x86  -Iinclude/arch/x86/guest  -Iinclude/debug  -Iinclude/public  -Iinclude/common  -Ibsp/include  -Ibsp/uefi/include/bsp  -Iboot/include -I. -m64 -nostdinc -nostdlib -gdwarf-2 -DASSEMBLER=1 -c arch/x86/vmx_asm.S -o build-test/hypervisor/arch/x86/vmx_asm.o
[ ! -e build-test/hypervisor/arch/x86/retpoline-thunk.o ] && mkdir -p build-test/hypervisor/arch/x86/; \
cc  -Iinclude  -Iinclude/lib  -Iinclude/lib/crypto  -Iinclude/common  -Iinclude/arch/x86  -Iinclude/arch/x86/guest  -Iinclude/debug  -Iinclude/public  -Iinclude/common  -Ibsp/include  -Ibsp/uefi/include/bsp  -Iboot/include -I. -m64 -nostdinc -nostdlib -gdwarf-2 -DASSEMBLER=1 -c arch/x86/retpoline-thunk.S -o build-test/hypervisor/arch/x86/retpoline-thunk.o
cc -E -x c  -Iinclude  -Iinclude/lib  -Iinclude/lib/crypto  -Iinclude/common  -Iinclude/arch/x86  -Iinclude/arch/x86/guest  -Iinclude/debug  -Iinclude/public  -Iinclude/common  -Ibsp/include  -Ibsp/uefi/include/bsp  -Iboot/include bsp/ld/link_ram.ld.in | grep -v '^#' > build-test/hypervisor/link_ram.ld
cc -Wl,-Map=build-test/hypervisor/acrn.map -o build-test/hypervisor/acrn.out -Wl,--gc-sections -static -nostartfiles -nostdlib -Wl,-n,-z,max-page-size=0x1000 -Wl,-z,noexecstack  -Tbuild-test/hypervisor/link_ram.ld build-test/hypervisor/boot/acpi.o build-test/hypervisor/boot/dmar_parse.o build-test/hypervisor/boot/multiboot.o build-test/hypervisor/arch/x86/ioapic.o build-test/hypervisor/arch/x86/intr_lapic.o build-test/hypervisor/arch/x86/cpu.o build-test/hypervisor/arch/x86/softirq.o build-test/hypervisor/arch/x86/cpuid.o build-test/hypervisor/arch/x86/mmu.o build-test/hypervisor/arch/x86/notify.o build-test/hypervisor/arch/x86/intr_main.o build-test/hypervisor/arch/x86/vtd.o build-test/hypervisor/arch/x86/gdt.o build-test/hypervisor/arch/x86/irq.o build-test/hypervisor/arch/x86/timer.o build-test/hypervisor/arch/x86/ept.o build-test/hypervisor/arch/x86/io.o build-test/hypervisor/arch/x86/interrupt.o build-test/hypervisor/arch/x86/vmexit.o build-test/hypervisor/arch/x86/vmx.o build-test/hypervisor/arch/x86/assign.o build-test/hypervisor/arch/x86/trusty.o build-test/hypervisor/arch/x86/cpu_state_tbl.o build-test/hypervisor/arch/x86/guest/vcpu.o build-test/hypervisor/arch/x86/guest/vm.o build-test/hypervisor/arch/x86/guest/instr_emul_wrapper.o build-test/hypervisor/arch/x86/guest/vlapic.o build-test/hypervisor/arch/x86/guest/guest.o build-test/hypervisor/arch/x86/guest/vmcall.o build-test/hypervisor/arch/x86/guest/vpic.o build-test/hypervisor/arch/x86/guest/vmsr.o build-test/hypervisor/arch/x86/guest/vioapic.o build-test/hypervisor/arch/x86/guest/instr_emul.o build-test/hypervisor/arch/x86/guest/ucode.o build-test/hypervisor/arch/x86/guest/pm.o build-test/hypervisor/arch/x86/debug/reboot.o build-test/hypervisor/lib/spinlock.o build-test/hypervisor/lib/udelay.o build-test/hypervisor/lib/mdelay.o build-test/hypervisor/lib/div.o build-test/hypervisor/lib/string.o build-test/hypervisor/lib/memory.o build-test/hypervisor/lib/crypto/hkdf.o build-test/hypervisor/lib/sprintf.o build-test/hypervisor/common/hv_main.o build-test/hypervisor/common/hypercall.o build-test/hypervisor/common/trusty_hypercall.o build-test/hypervisor/common/schedule.o build-test/hypervisor/common/vm_load.o build-test/hypervisor/common/io_request.o build-test/hypervisor/common/stack_protector.o build-test/hypervisor/bsp/uefi/vm_description.o build-test/hypervisor/bsp/uefi/uefi.o build-test/hypervisor/bsp/uefi/cmdline.o build-test/hypervisor/debug/printf.o build-test/hypervisor/debug/shell_public.o build-test/hypervisor/debug/console.o build-test/hypervisor/debug/dump.o build-test/hypervisor/debug/logmsg.o build-test/hypervisor/debug/shell_internal.o build-test/hypervisor/debug/vuart.o build-test/hypervisor/debug/sbuf.o build-test/hypervisor/debug/uart16550.o build-test/hypervisor/debug/serial.o build-test/hypervisor/arch/x86/cpu_secondary.o build-test/hypervisor/arch/x86/cpu_primary.o build-test/hypervisor/arch/x86/idt.o build-test/hypervisor/arch/x86/vmx_asm.o build-test/hypervisor/arch/x86/retpoline-thunk.o
objcopy -S --section-alignment=0x1000 -O elf32-i386 build-test/hypervisor/acrn.out build-test/hypervisor/acrn.32.out
objcopy -O binary build-test/hypervisor/acrn.out build-test/hypervisor/acrn.bin
echo "building hypervisor as EFI executable..."
building hypervisor as EFI executable...
make -C bsp/uefi/efi HV_OBJDIR=build-test/hypervisor RELEASE=0
make[2]: Entering directory '/home/gvancuts/work/acrn/acrn-hypervisor/hypervisor/bsp/uefi/efi'
[ ! -e build-test/hypervisor/bsp/uefi/efi/boot.o ] && mkdir -p build-test/hypervisor/bsp/uefi/efi/; \
cc  -I../include/bsp/ -I. -c -I. -I.. -I/usr/include/efi -I/usr/include/efi/x86_64 -DEFI_FUNCTION_WRAPPER -fPIC -fshort-wchar -ffreestanding -Wall -I../fs/ -Dx86_64 -O2 -mno-mmx -mno-sse -mno-sse2 -mno-80387 -mno-fp-ret-in-387 -mno-red-zone  boot.c -o build-test/hypervisor/bsp/uefi/efi/boot.o
[ ! -e build-test/hypervisor/bsp/uefi/efi/pe.o ] && mkdir -p build-test/hypervisor/bsp/uefi/efi/; \
cc  -I../include/bsp/ -I. -c -I. -I.. -I/usr/include/efi -I/usr/include/efi/x86_64 -DEFI_FUNCTION_WRAPPER -fPIC -fshort-wchar -ffreestanding -Wall -I../fs/ -Dx86_64 -O2 -mno-mmx -mno-sse -mno-sse2 -mno-80387 -mno-fp-ret-in-387 -mno-red-zone  pe.c -o build-test/hypervisor/bsp/uefi/efi/pe.o
[ ! -e build-test/hypervisor/bsp/uefi/efi/malloc.o ] && mkdir -p build-test/hypervisor/bsp/uefi/efi/; \
cc  -I../include/bsp/ -I. -c -I. -I.. -I/usr/include/efi -I/usr/include/efi/x86_64 -DEFI_FUNCTION_WRAPPER -fPIC -fshort-wchar -ffreestanding -Wall -I../fs/ -Dx86_64 -O2 -mno-mmx -mno-sse -mno-sse2 -mno-80387 -mno-fp-ret-in-387 -mno-red-zone  malloc.c -o build-test/hypervisor/bsp/uefi/efi/malloc.o
ld -T /usr/lib64/gnuefi/elf_x86_64_efi.lds -Bsymbolic -shared -nostdlib -znocombreloc -L/usr/lib64 /usr/lib64/gnuefi/crt0-efi-x86_64.o -o build-test/hypervisor/bsp/uefi/efi/boot.so build-test/hypervisor/bsp/uefi/efi/boot.o build-test/hypervisor/bsp/uefi/efi/pe.o build-test/hypervisor/bsp/uefi/efi/malloc.o  -lgnuefi -lefi /usr/lib/gcc/x86_64-redhat-linux/7/libgcc.a
objcopy -j .text -j .sdata -j .data -j .dynamic -j .dynsym -j .rel \
	-j .rela -j .reloc --target=efi-app-x86-64 build-test/hypervisor/bsp/uefi/efi/boot.so build-test/hypervisor/bsp/uefi/efi/boot.efi
objcopy  --add-section .hv="build-test/hypervisor/acrn.bin"  --change-section-vma .hv=0x6e000 --set-section-flags .hv=alloc,data,contents,load  --section-alignment 0x1000 build-test/hypervisor/bsp/uefi/efi/boot.efi  build-test/hypervisor/acrn.efi
objcopy: cannot open: build-test/hypervisor/acrn.bin: No such file or directory
make[2]: *** [Makefile:92: all] Error 1
make[2]: Leaving directory '/home/gvancuts/work/acrn/acrn-hypervisor/hypervisor/bsp/uefi/efi'
make[1]: *** [Makefile:179: efi] Error 2
make[1]: Leaving directory '/home/gvancuts/work/acrn/acrn-hypervisor/hypervisor'
make: *** [Makefile:17: hypervisor] Error 2
```

Additional data: building for `PLATFORM=sbl` goes further (i.e. the hypervisor is successfully built) but it fails at building the tools subdirectory. Here is the tail end of the output:
```
[ ! -e build-test/devicemodel/arch/x86/pm.o ] && mkdir -p build-test/devicemodel/arch/x86/; \
cc -g -O0 -std=gnu11 -D_GNU_SOURCE -DNO_OPENSSL -m64 -Wall -ffunction-sections -Werror -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -fno-strict-aliasing -I/home/gvancuts/work/acrn/acrn-hypervisor/devicemodel/include -I/home/gvancuts/work/acrn/acrn-hypervisor/devicemodel/include/public -fstack-protector-strong -c arch/x86/pm.c -o build-test/devicemodel/arch/x86/pm.o
cc -o build-test/devicemodel/acrn-dm -g -O0 -std=gnu11 -D_GNU_SOURCE -DNO_OPENSSL -m64 -Wall -ffunction-sections -Werror -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -fno-strict-aliasing -I/home/gvancuts/work/acrn/acrn-hypervisor/devicemodel/include -I/home/gvancuts/work/acrn/acrn-hypervisor/devicemodel/include/public -fstack-protector-strong -Wl,-z,noexecstack -Wl,-z,relro,-z,now build-test/devicemodel/hw/block_if.o build-test/devicemodel/hw/usb_core.o build-test/devicemodel/hw/uart_core.o build-test/devicemodel/hw/pci/virtio/virtio.o build-test/devicemodel/hw/pci/virtio/virtio_kernel.o build-test/devicemodel/hw/platform/usb_mouse.o build-test/devicemodel/hw/platform/atkbdc.o build-test/devicemodel/hw/platform/ps2mouse.o build-test/devicemodel/hw/platform/rtc.o build-test/devicemodel/hw/platform/ps2kbd.o build-test/devicemodel/hw/platform/ioapic.o build-test/devicemodel/hw/platform/cmos_io.o build-test/devicemodel/hw/platform/ioc.o build-test/devicemodel/hw/platform/ioc_cbc.o build-test/devicemodel/hw/platform/acpi/acpi.o build-test/devicemodel/hw/platform/acpi/acpi_pm.o build-test/devicemodel/hw/pci/wdt_i6300esb.o build-test/devicemodel/hw/pci/lpc.o build-test/devicemodel/hw/pci/xhci.o build-test/devicemodel/hw/pci/core.o build-test/devicemodel/hw/pci/virtio/virtio_console.o build-test/devicemodel/hw/pci/virtio/virtio_block.o build-test/devicemodel/hw/pci/virtio/virtio_input.o build-test/devicemodel/hw/pci/ahci.o build-test/devicemodel/hw/pci/hostbridge.o build-test/devicemodel/hw/pci/passthrough.o build-test/devicemodel/hw/pci/virtio/virtio_net.o build-test/devicemodel/hw/pci/virtio/virtio_rnd.o build-test/devicemodel/hw/pci/virtio/virtio_hyper_dmabuf.o build-test/devicemodel/hw/pci/virtio/virtio_heci.o build-test/devicemodel/hw/pci/irq.o build-test/devicemodel/hw/pci/uart.o build-test/devicemodel/core/monitor.o build-test/devicemodel/core/sw_load_common.o build-test/devicemodel/core/sw_load_bzimage.o build-test/devicemodel/core/sw_load_vsbl.o build-test/devicemodel/core/smbiostbl.o build-test/devicemodel/core/mevent.o build-test/devicemodel/core/gc.o build-test/devicemodel/core/console.o build-test/devicemodel/core/inout.o build-test/devicemodel/core/mem.o build-test/devicemodel/core/post.o build-test/devicemodel/core/consport.o build-test/devicemodel/core/vmmapi.o build-test/devicemodel/core/mptbl.o build-test/devicemodel/core/main.o build-test/devicemodel/core/hugetlb.o build-test/devicemodel/arch/x86/pm.o -lrt -lpthread -lcrypto -lpciaccess -lz -luuid
make[1]: Leaving directory '/home/gvancuts/work/acrn/acrn-hypervisor/devicemodel'
mkdir -p build-test/tools
make -C /home/gvancuts/work/acrn/acrn-hypervisor/tools/acrnlog OUT_DIR=build-test/tools
make[1]: Entering directory '/home/gvancuts/work/acrn/acrn-hypervisor/tools/acrnlog'
cc -g acrnlog.c -o build-test/tools/acrnlog -lpthread
/usr/bin/ld: cannot open output file build-test/tools/acrnlog: No such file or directory
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:5: all] Error 1
make[1]: Leaving directory '/home/gvancuts/work/acrn/acrn-hypervisor/tools/acrnlog'
make: *** [Makefile:25: tools] Error 2
```

--- a/Makefile
+++ b/Makefile
@@ -4,7 +4,8 @@ T := $(CURDIR)
 PLATFORM ?= uefi
 RELEASE ?= 0
 
-ROOT_OUT := $(T)/build
+BUILD_DIR := build
+ROOT_OUT := $(T)/$(BUILD_DIR)
 HV_OUT := $(ROOT_OUT)/hypervisor
 DM_OUT := $(ROOT_OUT)/devicemodel
 TOOLS_OUT := $(ROOT_OUT)/tools

